src/http-adapter/index.js

```diff
@@ -0,0 +1,228 @@
+  constructor (options = {}) {
+    this.urlPrefix = options.urlPrefix
+    this.serializeRequests = false
+    this._outstandingRequests = new Set()
+    this.requestInterceptor = options.requestInterceptor || (() => {})
+    this.responseInterceptor = options.responseInterceptor || (() => {})
+  }
+
+  /**
+   * Parses a JSON:API resource to a restful url.
+   * EX. resource with type 'cat' and id '1' with a urlPrefix 'animal' will return '/animal/cat/1/`
+   * @param {string} type - A JSON:API resource type. See https://jsonapi.org/format/#document-resource-objects
+   * @param {string} id - A JSON:API resource id. See https://jsonapi.org/format/#document-resource-objects
+   * @returns {string} - the built url containing the urlPrefix.
+   */
+  buildUrl (type, id) {
+    let idPath = `/${id}`
+    let typePath = `/${type}`
+    let path = this.urlPrefix || ''
+
+    path += typePath
+
+    if (id != null) {
+      path += idPath
+    }
+
+    return `${path}/`
+  }
+
+  /**
+   * Get entity by link.
+   * @private
+   * @param {string} link - Link to entity.
+   * @param {Object} query - query params for a given request in a key/value pair format
+   * @returns {Promise} Promise for fetched data.
+   */
+  get (link, query) {
+    return this._http(HTTP_METHOD.GET, link, query)
+  }
+
+  /**
+   * Create entity.
+   * @private
+   * @param {string} link - Entity url.
+   * @param {object} payload - Data to create entity with.
+   * @returns {Promise} Promise for created data.
+   */
+  create (link, payload) {
+    return this._http(HTTP_METHOD.POST, link, payload)
+  }
+
+  /**
+   * Update entity.
+   * @private
+   * @param {string} link - Entity url.
+   * @param {object} payload - Data to update entity with.
+   * @returns {Promise} Promise for updated data.
+   */
+  update (link, payload) {
+    return this._http(HTTP_METHOD.PATCH, link, payload)
+  }
+
+  /**
+   * Destroy entity.
+   * @private
+   * @param {string} link - Entity self link.
+   * @returns {Promise} Promise for destroy.
+   */
+  destroy (link) {
+    return this._http(HTTP_METHOD.DELETE, link)
+  }
+
+  /**
+   * Make a generic request that requires more options then CRUD methods provided.
+   * Does not set default Accept or Content-Type. This must be provided!
+   * @param {string} url - URL where the request is being sent.
+   * @param {Object} [options={}] - request options to be propagated to _makeRequest
+   * @param {string} options.url - the url of the request
+   * @param {HTTP_METHOD} options.method - the method of the request
+   * @param {Object} options.headers - A key/value object of headers to be sent with the request
+   * @param {Object} options.data - A key/value payload to be sent with the request. If method is GET, these
+   * options will be transformed into query paramters
+   * @param {boolean} options.isInternal - Whether a request is being sent to an internal server
+   * or not to invoke request/response interceptors
```

Comment:
Non-blocker: May concur with Rob that I'm not sure we need to call this `isInternal` instead of `useInterceptors`. The risk of not using `isInternal` is the user may not know we *need* to use interceptors for internal requests, but we can add that warning to the jsdoc.

One reason we may want to name it `useInterceptors` instead of `isInternal` is the current name couples interceptor use to whether `isInternal` is true. What if the user wants to use the interceptors, but the request is _not_ for our internal api server?

Reply:
@electrachong I 💯 agree. The only concern I have is for this line [here](https://github.com/groveco/backbone.store/pull/76/files#diff-528af5ead8558733b81e24e0441003e9R78), which I do not even believe is still applicable. Maybe we want to pass `isInternal=false` to `_makeRequest`  from `request`, but do not make it a part of the public API. An instead of `isInternal` for the public API, we use the option `useInterceptors`. What do you think? FYI @jessewagstaff

Reply:
I think it would make sense for `isInternal` to not be part of the public API, except for cases where we want to make a request to the internal api without going through backbone store. So maybe keep it as part of the public API? Or add a separate method for internal requests?

Edit: Or we make it the responsibility of the app to toggle `dataType` to something else in an interceptor if we don't want it to be `text` for third-party requests? Given that we automatically parse the response into JSON, it seems like we'd want to keep the dataType to text even for third-party requests. If ajax automatically parses the response into json, it seems like the adapter might throw a syntax error trying to parse the response again.